_posts/2019-02-05-configurable-builds-part-1.md

```diff
@@ -0,0 +1,284 @@
+single project.
+
+One of the reasons
+
+    $ bazel build //my:binary --cpu=arm --crosstool_top=//my:custom_toolchain --define MYFEATURE=1
+
+doesn't work well is that `--cpu` and `--crosstool_top` are C++-specific. What
+if you want to build Java libraries with C++ dependencies for both Linux and
+Android? This impacts not just C++, but also the JDK and maybe even which
+support libraries you need.
+
+Not only are `--cpu` and `-crosstool_top` insufficiently expressive, but they
+might actively sabotoge you. Consider the following use of
+[select](https://docs.bazel.build/versions/master/configurable-attributes.html):
+
+```python
+java_library(
+    name = "supporting_library",
+	deps = select({
+	    ":android": [":extra_required_android_deps"],
+		"//conditions:default": []
+    }))
+```
+
+If you use `--cpu` and `crosstool_top` to define what `:android` means, what
+happens when an app supports a new Android phone with a different CPU?
+`:android` won't trigger because `--cpu` no longer matches, the app won't get
+required Android support libraries, and it will crash.
+
+### A Better Way
+
+The configurability's team's platform work, led by [@katre](https://github.com/katre),
+lets you declare exactly what you want in a way all rules understand. It
+doesn't even matter if no one's ever heard of your platform.
+
+With the new API, our example can be rewritten as:
+
+```python
+$ cat platforms/BUILD
+constraint_setting(name = "os")
+constraint_value(name = "android", constraint_setting = ":os")
+constraint_value(name = "linux", constraint_setting = ":os")
+
+constraint_setting(name = "device")
+constraint_value(name = "phone", constraint_setting = ":device")
+constraint_value(name = "tablet", constraint_setting = ":device")
+
+platform(name = "pixel3", constraint_values = [":android", ":phone"])
+```
+
+```python
+$ cat helpers/BUILD
+config_setting(name = "android", constraint_values = "//platforms:android")
+java_library(
+    name = "supporting_library",
+    deps = select({
+	    ":android": [":extra_required_android_deps"],
+        "//conditions:default": []
+    }))
+```
+
+and built with:
+
+    $ bazel build //my:binary --platforms=//platforms:pixel3 --define MYFEATURE=1
+
+This is more concise and correct than before. Now it doesn't matter what the
+CPU is. Any platform with the `:android`
+[`constraint_value`](https://docs.bazel.build/versions/master/be/platform.html#constraint_value)
+is an Android platform. So it's easy to express *exactly* what you want and
+have `select`, rules, and toolchains understand it the same way. `MYFEATURE=1`,
+which isn't a platform property, remains as before.
+
+Rules understand `:android` by declaring [toolchain types](https://docs.bazel.build/versions/master/toolchains.html#writing-rules-that-use-toolchains).
+For example, `java_binary` can be defined in Starlark as
+
+```python
+java_binary = rule(..., toolchains = ["//toolchain_types:java"])
```

Comment:
Check with @katre, but I think the convention is `//something_about_java:toolchain_type`, i.e. literally a target named "toolchain_type".

Reply:
Done.